src/new/wallet.js

```diff
@@ -0,0 +1,441 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import EventEmitter from 'events';
+import networkInstance from '../network';
+import wallet from '../wallet';
+import { GAP_LIMIT, HATHOR_TOKEN_CONFIG, DEFAULT_SERVER } from '../constants';
+import transaction from '../transaction';
+import tokens from '../tokens';
+import version from '../version';
+import ws from '../WebSocketHandler';
+import storage from '../storage';
+import MemoryStore from '../memory_store';
+import walletApi from '../api/wallet';
+
+/**
+ * This is a Wallet that is supposed to be simple to be used by a third-party app.
+ *
+ * This class handles all the details of syncing, including receiving the same transaction
+ * multiple times from the server. It also keeps the balance of the tokens updated.
+ *
+ * It has the following states:
+ * - CLOSED: When it is disconnected from the server.
+ * - CONNECTING: When it is connecting to the server.
+ * - SYNCING: When it has connected and is syncing the transaction history.
+ * - READY: When it is ready to be used.
+ *
+ * You can subscribe for the following events:
+ * - state: Fired when the state of the Wallet changes.
+ * - new-tx: Fired when a new tx arrives.
+ * - update-tx: Fired when a known tx is updated. Usually, it happens when one of its outputs is spent.
+ * - more-addresses-loaded: Fired when loading the history of transactions. It is fired multiple times,
+ *                          one for each request sent to the server.
+ **/
+class Wallet extends EventEmitter {
+  constructor({ network, server, seed, tokenUid }) {
+    super();
+
+    networkInstance.setNetwork(network);
+    this.network = network;
+
+    this.state = Wallet.CLOSED;
+    this.serverInfo = null;
+
+    this.onConnectionChange = this.onConnectionChange.bind(this);
+    this.handleWebsocketMsg = this.handleWebsocketMsg.bind(this);
+    this.onAddressesLoaded = this.onAddressesLoaded.bind(this);
+
+    this.seed = seed;
+    this.server = server;
+    if (!this.server) {
+      // We set the default server of the lib
+      this.server = DEFAULT_SERVER;
+
+    }
+
+    // tokenUid is optional so we can get the token of the wallet
+    this.token = null;
+    this.tokenUid = tokenUid;
+
+
+    this.passphrase = '';
+    // XXX Update it so we don't have fixed pin/password
+    this.pinCode = '123';
```

Comment:
Maybe we can create an issue to discuss what to do about the pin and password. I'm not sure the lib should be handling them. It looks like something that should be one layer above. Whoever uses the lib decides what safety mechanism he wants to implement.

Reply:
I agree. Current implementation requires a pinCode and passphrase to store sensitive data. We should refactor the internals to be able to change it here.

Reply:
I agree also, it should be in the refactor of the lib issue. But I think we should have this disabled but available as a default in the lib for apps to use it. So we also can use in both desktop and mobile wallets.

Like our memory storage, that we have a default and, if the user wants to change, they are able to do it.

Reply:
Added to the lib issue